pennylane_sf/vgbs.py

```diff
@@ -0,0 +1,142 @@
+
+
+class StrawberryFieldsVGBS(StrawberryFieldsSimulator):
+    r"""TODO
+    """
+    name = 'Strawberry Fields variational GBS PennyLane plugin'
+    short_name = 'strawberryfields.vgbs'
+
+    _operation_map = {
+        'Displacement': Dgate,
+        "GraphTrain": GraphEmbed,
+    }
+
+    _observable_map = {
+        'Identity': identity
+    }
+
+    _circuits = {}
+
+    _capabilities = {"model": "cv", "provides_jacobian": True}
+
+    def __init__(self, wires, *, analytic=True, cutoff_dim, backend="gaussian", shots=1000, hbar=2):
+        if not analytic and backend != "gaussian":
+            raise ValueError("Only the Gaussian backend is supported in non-analytic mode.")
+
+        super().__init__(wires, analytic=analytic, shots=shots, hbar=hbar)
+        self.cutoff = cutoff_dim
+        self.backend = backend
+
+    def apply(self, operation, wires, par):
+        """TODO
+        """
+        weights, A, n_mean = par
+        A = A * rescale(A, n_mean)
+        W = np.diag(np.sqrt(weights))
+        WAW = W @ A @ W
+
+        singular_values = np.linalg.svd(WAW, compute_uv=False)
+        n_mean_WAW = np.sum(singular_values ** 2 / (1 - singular_values ** 2))
+
+        op = self._operation_map[operation](WAW, mean_photon_per_mode=n_mean_WAW / len(A))
+        op | [self.q[i] for i in wires] #pylint: disable=pointless-statement
+
+        if not self.analytic:
+            MeasureFock() | [self.q[i] for i in wires]
+
+        self.weights = weights
+
+    def pre_measure(self):
+        self.eng = sf.Engine(self.backend, backend_options={"cutoff_dim": self.cutoff})
+
+        if self.analytic:
+            results = self.eng.run(self.prog)
+        else:
+            # NOTE: currently, only the gaussian backend supports shots > 1
+            results = self.eng.run(self.prog, shots=self.shots)
+
+        self.state = results.state
+        self.samples = results.samples
+
+    def probability(self, wires=None):
+        if self.analytic:
+            # compute the fock probabilities analytically
+            # from the state representation
+            return super().probability(wires=wires)
+
+        # compute the fock probabilities from samples
+        N = len(wires) if wires is not None else len(self.num_wires)
+        probs = all_fock_probs_pnr(self.samples)
+        ind = np.indices([self.cutoff] * N).reshape(N, -1).T
+        probs = OrderedDict((tuple(k), v) for k, v in zip(ind, probs))
+        return probs
+
+    def jacobian(self, operations, observables, variable_deps):
+        # NOTE: potentially could provide caching here, to avoid recomputing
+        # samples/probabilities.
+        self.reset()
+        prob = np.squeeze(self.execute(operations, observables, parameters=variable_deps))
+
+        # Compute and return the jacobian here.
+        # returned jacobian matrix should be of size (output_length, num_params)
+        jac = np.zeros([len(prob), len(self.weights)])
+        return jac
```

Comment:
@trbromley, I've added scaffolding here for the jacobian method, let me know if you have any questions! I wasn't entirely sure of how to calculate the jacobian in this case.

You can make any of the variables in `apply()` attributes if you need to re-use them here.